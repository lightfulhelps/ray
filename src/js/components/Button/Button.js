// @flow
import * as React from 'react';
import { color, number } from '@storybook/addon-knobs';
import classNames from 'classnames';
import Icon from '../Icon/Icon';

type Props = {
  children: React.Element<any>,
  className?: string,
  isBlock?: boolean,
  isDisabled?: boolean,
  isOutline?: boolean,
  onClick?: () => void,
  size?: 'lg' | 'md' | 'sm',
  tag?: string,
  theme?: string,
  icon?: string,
  iconPosition?: string,
  isDropdown?: boolean,
};

const Button = ({
  children,
  label,
  className,
  theme = 'primary',
  isBlock,
  isDisabled,
  isOutline,
  onClick = () => {},
  size,
  tag: Tag = 'button',
  icon,
  iconPosition,
  isDropdown,
}: Props) => {
  const classes = classNames(
    className,
    'btn',
    `btn${isOutline ? '-outline' : ''}-${theme}`,
    size ? `btn-${size}` : false,
    { 'btn-block': isBlock },
    { disabled: isDisabled },
    { 'btn-icon': icon && !label},
    { [`btn-icon-${iconPosition}`]: iconPosition && label }
  );

  return (
    <Tag 
      className={classes} 
      onClick={onClick}
      data-toggle={isDropdown && "dropdown"}
      aria-haspopup={isDropdown && "true"}
      aria-expanded={isDropdown && "false"}
    >
      {label && (
        <span>{label}</span>
      )}
      {icon && (
        <Icon
          name={icon}
          size={number('Size', 24)}
          color={color('Color', '#adb5bd')}
          hoverColor={color('Hover Color', '#212529')}
        />
      )}
    </Tag>
  );
};

export default Button;
