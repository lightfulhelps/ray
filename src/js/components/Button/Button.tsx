import * as React from 'react';
import classNames from 'classnames';
import Icon from '../Icon/Icon';
import { IconNameType } from '../Icon/icons';

// Size is an attribute of the input html element so there was a conflict with our size props.
// Hence the need to omit this attribute
type ButtonAttributes = Omit<React.HTMLProps<HTMLButtonElement>, 'size'>;

export type Props = {
  animateGradient?: boolean;
  children?: React.ReactNode;
  className?: string;
  forceSolidColor?: boolean;
  icon?: IconNameType;
  iconPosition?: 'left' | 'right';
  iconSize?: string;
  iconTheme?: string;
  isBlock?: boolean;
  isDisabled?: boolean;
  isOutline?: boolean;
  loading?: boolean;
  onClick?: (event: React.MouseEvent) => void;
  size?: 'lg' | 'md' | 'sm';
  tag?: keyof JSX.IntrinsicElements | typeof React.Component;
  theme?: string;
  type?: 'button' | 'submit' | 'reset';
  [key: string]: any;
};

const gradientThemes = [
  'primary',
  'secondary',
  'tertiary-red',
  'tertiary-green',
  'tertiary-black',
  'academy',
  'success',
  'neutral',
];
const Button: React.FC<ButtonAttributes & Props> = ({
  children,
  className,
  forceSolidColor = false,
  icon,
  iconPosition = 'left',
  iconSize = '1em',
  iconTheme,
  isBlock,
  isDisabled,
  isOutline,
  loading = false,
  onClick,
  size,
  tag: Tag = 'button',
  theme = 'primary',
  type = 'button',
  animateGradient = false,
  ...other
}: Props) => {
  const classes = classNames(
    className,
    'btn',
    { [`btn${isOutline ? '-outline' : ''}-gradient-${theme}`]: !forceSolidColor },
    `btn${isOutline ? '-outline' : ''}-${theme}`, // backup
    size ? `btn-${size}` : false,
    { disabled: isDisabled },
    { 'btn-icon': icon && !children },
    { [`btn-icon-${iconPosition}`]: icon && children },
    { loading },
    { 'animate-gradient': animateGradient }
  );

  const handleClick = (e: React.MouseEvent): void => {
    if (isDisabled) {
      e.preventDefault();
      return;
    }

    if (typeof onClick === 'function') onClick(e);
  };

  // Workaround to optionally add type attribute since this one is only available on button Jsx element
  const optionalProps: { [key: string]: any } = {};
  optionalProps.type = Tag === 'button' ? type : undefined;

  const buttonComponent = (
    <Tag
      {...other}
      {...optionalProps}
      className={classes}
      onClick={handleClick}
      disabled={isDisabled}
    >
      <span>
        {children}
        {icon && (
          <Icon
            name={icon}
            size={iconSize}
            theme={iconTheme || (isOutline && theme)}
            isGradient={gradientThemes.includes(theme) && !forceSolidColor}
          />
        )}
      </span>
    </Tag>
  );

  if (isBlock) return <div className="d-grid">{buttonComponent}</div>;
  return buttonComponent;
};

export default Button;
