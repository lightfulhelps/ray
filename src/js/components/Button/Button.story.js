import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, text, select, boolean } from '@storybook/addon-knobs';
import Button from './Button';

import '../../../scss/ray.scss';

const stories = storiesOf('Button', module);

const themes = {
  primary: 'Primary',
  secondary: 'Secondary',
  success: 'Success',
  danger: 'Danger',
  warning: 'Warning',
  info: 'Info',
  light: 'Light',
  dark: 'Dark',
};

const sizes = {
  lg: 'Large',
  md: 'Medium',
  sm: 'Small',
};

stories.addDecorator(withKnobs);

stories.add('Default', () => (
  <Button
    onClick={action('clicked')}
    color={select('Theme', themes, 'primary')}
    outline={boolean('Outline', false)}
    disabled={boolean('Disabled', false)}
    size={select('Size', sizes, 'md')}
  >
    {text('Text', 'Click Me')}
  </Button>
));
