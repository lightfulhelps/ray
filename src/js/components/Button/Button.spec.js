import React from 'react';
import { shallow } from 'enzyme';
import merge from 'lodash/merge';
import Button from './Button';

const setup = (overrides = {}) => {
  const props = merge(
    {
      children: 'Click Me',
      onClick: jest.fn(),
    },
    overrides
  );
  const wrapper = shallow(<Button {...props} />);

  return {
    wrapper,
    props,
  };
};

describe('<Button />', () => {
  it('should render', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should render with an icon', () => {
    const { wrapper } = setup({ icon: 'create' });

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle the iconSize props', () => {
    const { wrapper } = setup({ icon: 'create', iconSize: 24 });

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle the iconTheme props', () => {
    const { wrapper } = setup({ icon: 'create', iconTheme: 'primary' });

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle the children prop', () => {
    const { wrapper } = setup();

    expect(wrapper.children().text()).toEqual('Click Me');
  });

  it('should handle the className prop', () => {
    const { wrapper } = setup();

    wrapper.setProps({ className: 'custom' });

    expect(wrapper.hasClass('btn')).toBe(true);
    expect(wrapper.hasClass('custom')).toBe(true);
  });

  it('should handle the isBlock prop', () => {
    const { wrapper } = setup();

    expect(wrapper.hasClass('btn-block')).toBe(false);

    wrapper.setProps({ isBlock: true });

    expect(wrapper.hasClass('btn-block')).toBe(true);
  });

  it('should handle the isDisabled prop', () => {
    const { wrapper } = setup();

    expect(wrapper.hasClass('disabled')).toBe(false);

    wrapper.setProps({ isDisabled: true });

    expect(wrapper.hasClass('disabled')).toBe(true);
  });

  it('should handle the isOutline prop', () => {
    const { wrapper } = setup();

    expect(wrapper.hasClass('btn-outline-primary')).toBe(false);
    expect(wrapper.hasClass('btn-primary')).toBe(true);

    wrapper.setProps({ isOutline: true });

    expect(wrapper.hasClass('btn-outline-primary')).toBe(true);
    expect(wrapper.hasClass('btn-primary')).toBe(false);
  });

  it('should handle the size prop', () => {
    const { wrapper } = setup();

    wrapper.setProps({ size: 'lg' });

    expect(wrapper.hasClass('btn-lg')).toBe(true);

    wrapper.setProps({ size: 'sm' });

    expect(wrapper.hasClass('btn-sm')).toBe(true);
  });

  it('should handle the tag prop', () => {
    const { wrapper } = setup();

    expect(wrapper.type()).toBe('button');

    wrapper.setProps({ tag: 'a' });

    expect(wrapper.type()).toBe('a');

    wrapper.setProps({ tag: 'div' });

    expect(wrapper.type()).toBe('div');
  });

  it('should handle the theme prop', () => {
    const { wrapper } = setup();

    expect(wrapper.hasClass('btn-primary')).toBe(true);

    wrapper.setProps({ theme: 'secondary' });

    expect(wrapper.hasClass('btn-primary')).toBe(false);

    expect(wrapper.hasClass('btn-secondary')).toBe(true);
  });

  it('should default the type prop to button if Tag is button', () => {
    const { wrapper } = setup();

    wrapper.setProps({ tag: 'button' });

    expect(wrapper.prop('type')).toEqual('button');

    wrapper.setProps({ type: 'submit' });

    expect(wrapper.prop('type')).toEqual('submit');

    wrapper.setProps({ tag: 'a' });

    expect(wrapper.prop('type')).toEqual(undefined);
  });

  it('should handle the iconPosition prop', () => {
    const { wrapper } = setup({ icon: 'create' });

    expect(wrapper.hasClass('btn-icon-left')).toBe(true);

    wrapper.setProps({ iconPosition: 'right' });

    expect(wrapper.hasClass('btn-icon-right')).toBe(true);

    wrapper.setProps({ iconPosition: 'left' });

    expect(wrapper.hasClass('btn-icon-left')).toBe(true);
  });

  it('should pass through other props', () => {
    const { wrapper } = setup({ tabIndex: 1, id: 'test' });

    expect(wrapper.prop('tabIndex')).toEqual(1);
    expect(wrapper.prop('id')).toEqual('test');
  });

  describe('onClick', () => {
    it('should call onClick with data provided', () => {
      const { wrapper, props } = setup();

      wrapper.simulate('click', { foo: 'bar' });

      expect(props.onClick).toHaveBeenCalledTimes(1);
      expect(props.onClick).toHaveBeenCalledWith({ foo: 'bar' });
    });

    it('should preventDefault if isDisabled', () => {
      const preventDefault = jest.fn();
      const { wrapper } = setup();

      wrapper.simulate('click');

      expect(preventDefault).not.toHaveBeenCalled();

      wrapper.setProps({ isDisabled: true });

      wrapper.simulate('click', { preventDefault });

      expect(preventDefault).toHaveBeenCalledTimes(1);
    });

    it('should not error if onClick is not a function', () => {
      const { wrapper } = setup({ onClick: 'foo' });

      expect(() => {
        wrapper.simulate('click');
      }).not.toThrow();
    });
  });
});
