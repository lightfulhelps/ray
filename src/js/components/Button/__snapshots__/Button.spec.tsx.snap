// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should handle the iconTheme props 1`] = `
<button
  className="btn btn-gradient-primary btn-primary btn-icon-left"
  onClick={[Function]}
  type="button"
>
  <span>
    Click Me
    <Icon
      isGradient={true}
      name="create"
      theme="primary"
    />
  </span>
</button>
`;

exports[`<Button /> should render 1`] = `
<button
  className="btn btn-gradient-primary btn-primary"
  onClick={[Function]}
  type="button"
>
  <span>
    Click Me
  </span>
</button>
`;

exports[`<Button /> should render with an icon 1`] = `
<button
  className="btn btn-gradient-primary btn-primary btn-icon-left"
  onClick={[Function]}
  type="button"
>
  <span>
    Click Me
    <Icon
      isGradient={true}
      name="create"
    />
  </span>
</button>
`;
